fix: enable gateway bootstrap when starting with zero connections

sanity opened the pull request:
Fixes #1863
Addresses #1870

## Problem

Gateways were unable to bootstrap when starting with zero connections, creating a deadlock:
- Connection maintenance task needs existing connections to query through
- `forward_conn()` requires `num_connections() > 0` to proceed  
- No connections could ever be established

This was likely broken during recent refactoring (per @iduartgomez).

## Solution

This PR implements a bootstrap mechanism for gateways while maintaining connection guarantee semantics:

1. **Enhanced logging** - Added comprehensive logging to track connection maintenance and bootstrap behavior
2. **Gateway bootstrap logic** - When a gateway has 0 connections and has already accepted an incoming connection via `should_accept()`, we allow the connection to proceed by returning `ConnectState::AwaitingConnectivity`
3. **Preserves semantics** - The `num_connections() == 0` check remains in place for non-bootstrap scenarios, maintaining @iduartgomez's requirement for connection stability guarantees
4. **Critical fix** - Corrected gateway detection logic to use actual `is_gateway` flag instead of incorrectly using `accepted == true` as a proxy (which would have triggered bootstrap for ANY node with 0 connections)

## Key Changes

- `ring/mod.rs`: 
  - Added detailed logging to `connection_maintenance()` task
  - Enhanced `acquire_new()` with bootstrap detection
  - Made task failures more visible (error level logging)

- `connect.rs`:
  - Added `is_gateway` field to `ForwardParams` struct
  - Special case for gateway bootstrap when `is_gateway && accepted && num_connections() == 0`
  - Creates proper `ConnectState::AwaitingConnectivity` to ensure connection gets added to ring
  - Maintains existing semantics for all other cases

- `handshake.rs`:
  - Added `is_gateway` field to `HandshakeHandler` struct
  - Passes `is_gateway` flag through to `forward_conn()`

- `p2p_protoc.rs`:
  - Passes `is_gateway` from `P2pBridge` to `HandshakeHandler`

## Test Results

### ✅ Unit Tests
```bash
cargo test --package freenet --lib ring
```
**Result**: All 9 tests passed

### ✅ Integration Tests  
```bash
cargo test --test connectivity
```
**Result**: Both connectivity tests passed (27.36s)
- `test_gateway_bootstrap` - Verified gateway can accept first connection
- `test_gateway_reconnection` - Verified peer can reconnect after disconnection

### ✅ Multi-Machine Test
```bash
cd ~/code/freenet/freenet-testing-tools/multi-machine-test
./run-test.sh local
```
**Result**: Connection layer successful ✅
- Gateway started successfully (72.29s)
- **Peer connected to gateway in 55.00s** ✅
- **Gateway bootstrap from 0 connections working correctly** ✅

*Note: Test ultimately failed in River chat layer (WebSocket disconnection) which is unrelated to the connection bootstrap fix.*

### Test Documentation

The most robust test is the **multi-machine-test** which:
- Deploys a gateway to a remote machine (vega.locut.us)
- Runs a local peer that connects to the remote gateway
- Tests real network conditions and multi-hop scenarios
- Validates gateway bootstrap from zero connections

This test confirms that the fix resolves issue #1863 - gateways can now successfully bootstrap and accept their first connection.

## Important Note

This maintains the architectural requirement that only "guaranteed" connections are used for forwarding (per @iduartgomez). The bootstrap case is a special exception where:
1. The node is actually a gateway (`is_gateway == true`)
2. The gateway has already decided to accept the connection via `should_accept()`
3. This ensures the connection meets stability requirements

[AI-assisted debugging and implementation]

sanity commented:
@claude Please critically review this PR relative to the Github Issues referenced in the description.

iduartgomez commented on crates/core/src/ring/mod.rs:
```diff
@@ -526,6 +574,20 @@ impl Ring {
+                    is_gateway
+                );
+
+                // BOOTSTRAP FIX: If we're a gateway with 0 connections,
```
```suggestion
                If we're a gateway with 0 connections,
```

iduartgomez commented:
@claude check ci for failures and address them

iduartgomez commented on crates/core/tests/connectivity.rs:
```diff
@@ -420,3 +420,302 @@ async fn test_basic_gateway_connectivity() -> TestResult {
+/// 2. Connect a first peer to the gateway
+/// 3. Connect a second peer and verify connectivity between all nodes
+#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
+async fn test_gateway_bootstrap_three_node_network() -> TestResult {
```
@claude in order to test connectivity simply query via the websocket to get all conenctions

iduartgomez commented:
@claude check the new unit tests failures and address those, keep in mind previous changes and the overall goal of this PR

iduartgomez commented:
@claude there is obviously somethign wrong either in the implementation or the test cause the new integration test is failing

@sanity if this gets fixed I think we can merge but not before, getting through via PR may be very cumbersome so may need some love in a proper local dev environment to get through claude...

iduartgomez commented:
> **Claude finished @iduartgomez's task** —— [View job](https://github.com/freenet/freenet-core/actions/runs/18126659571)
> ### Understanding the Gateway Connection Issue
> 
> **Todo List:**
> 
>     * [x]  Investigate original connection acceptance logic before the fix
> 
>     * [x]  Review the recent changes to understand what was modified
> 
>     * [x]  Check handshake_handler tests to see if acceptance was tested
> 
>     * [x]  Explain the distinction between accepting connections and using them for forwarding
> 
>     * [x]  Provide clear answer about why gateways weren't accepting connections below min threshold
> 
> 
> ## You're absolutely right to question this!
> 
> After investigating the code, I found that **gateways WERE already accepting connections when below minimum threshold** via `ConnectionManager::should_accept()`. The issue was a disconnect between two different parts of the system:
> ### 1. Connection Acceptance (`should_accept()`)
> 
> The `ConnectionManager::should_accept()` in `ring/connection_manager.rs:166` was already correctly returning `true` when below `min_connections`:
> 
> ```rust
> let accepted = if total_conn < self.min_connections {
>     tracing::debug!(%peer_id, "Accepted connection, below min connections");
>     true
> }
> ```
> 
> ### 2. Connection Forwarding (`forward_conn()`)
> 
> However, in `operations/connect.rs`, the `forward_conn()` function had a hard check that **blocked ALL connections when num_connections == 0**, regardless of whether the node was a gateway and had already accepted the connection:
> 
> ```rust
> // BEFORE the fix:
> if connection_manager.num_connections() == 0 {
>     if is_gateway && accepted {
>         // Only allowed bootstrap for exactly 0 connections
>         return Ok(Some(ConnectState::AwaitingConnectivity(...)));
>     } else {
>         // This would reject even if gateway with 1 connection and accepted==true!
>         return Ok(None);
>     }
> }
> ```
> 
> ### The Problem
> 
> This created the exact scenario you suspected:
> 
>     * Gateway with 1 connection calls `should_accept()` → returns `true` (because 1 < 25 min_connections)
> 
>     * Gateway reserves the connection slot
> 
>     * But then `forward_conn()` rejects it because it only allowed the bootstrap path for **exactly 0 connections**, not for 1-24 connections
> 
>     * Connection gets dropped despite being "accepted"
> 
> 
> ### The Fix
> 
> The fix (commit [8ef51ee](https://github.com/freenet/freenet-core/commit/8ef51ee937ee8a9f8e1a6ebc5f21372c0d39af2c)) aligned `forward_conn()` with `should_accept()` by allowing gateways to directly accept connections whenever they're below `min_connections`:
> 
> ```rust
> // AFTER the fix:
> if is_gateway && accepted && num_connections < connection_manager.min_connections {
>     // Gateway can accept directly when below threshold (0-24 connections)
>     return Ok(Some(ConnectState::AwaitingConnectivity(...)));
> }
> ```
> 
> ### Why wasn't this caught in tests?
> 
> The handshake tests only tested the extreme case (0 connections bootstrap) but not the intermediate case (1-24 connections). The new integration test exposed this by trying to connect a second peer when the gateway had 1 connection, which failed before the fix.
> 
> **In summary**: You're correct - gateways SHOULD have been accepting connections when below min threshold, and `should_accept()` was doing its job. The bug was that `forward_conn()` was incorrectly rejecting these already-accepted connections unless the gateway had exactly 0 connections. This was an implementation bug, not a design issue.
> 
> [View job](https://github.com/freenet/freenet-core/actions/runs/18127010445)

@claude what happens with regular peers?

shouldn't we address the 1 to MIN_CONNECTIONS cases in hadshake_handler tests?

iduartgomez commented:
@claude isn't the problem?

> Gateway calls forward_conn() → returns None, drops connection ❌
> Connection slot is reserved but never filled → deadlock

1. even if we CANNOT forward a connection we shouldn't be dropping an otherwise perfectly valid connection (!), is not this what we should be addressing?
2. if a connection slot was reserved, but it never gets filled, that reservation should be reversed

we seem to be addressing the wrong problem here

iduartgomez commented on crates/core/src/operations/connect.rs:
```diff
@@ -1025,8 +1026,27 @@ where
+            req_peer.clone(),
+            1, // Single check for direct connection
+        );
+        return Ok(Some(ConnectState::AwaitingConnectivity(connectivity_info)));
```
@claude won't the issue now be that even though we will accept the connection, we won't forward it if it was possible? or will the forwarding resume anyway later on?

or should we attempt to forward and regardless of the result accept the connection?

It may be necessary to add new intermediate states in the op state machine

iduartgomez commented:
@claude this is looking better, but fix all the compile errors, lint warnings and formatting

iduartgomez commented on crates/core/tests/connectivity.rs:
```diff
@@ -420,3 +420,372 @@ async fn test_basic_gateway_connectivity() -> TestResult {
+
+        // Query gateway for its connections
+        tracing::info!("Querying gateway for connected peers...");
+        client_gw
```
@claude for this test to not be flaky, you may have to do the different NodeQueries multiple times in a loop while all peers connect to each other, and fail after a few seconds if they fail not to connect in time, other way you will have a race condition in the PR

additionally you must check that the ConnectedPeers of each node contains the 2 other peers

sanity commented:
## Investigation Results: test_three_node_network_connectivity Failure

I've systematically investigated why this test has been failing since it was added (commit 48ed3ffa).

### Bugs Found

**1. Location Mismatch Bug (Fixed)**
The gateway's location was being generated twice with different random values:
- `gateway_config.location = RNG.random()` (line 471)  
- `gateway_info.location = RNG.random()` (line 492)

This caused peers to have incorrect routing information about the gateway. Fixed by generating the location once and reusing it (commit in progress).

**2. Root Cause: Test Design vs Configuration Mismatch**

The test expects **full mesh topology** (each of 3 nodes connected to both others), but peers are configured with:
- `public_port: None`
- `network_port: None`

Without public ports, peers cannot accept incoming connections. They can only initiate outbound connections to the gateway.

**Current behavior:**
- Gateway: 2 connections ✓
- Peer1: 1 connection (gateway only)
- Peer2: 1 connection (gateway only)

This is a **star topology**, not a mesh.

### Question

@iduartgomez: Should non-gateway peers be able to form direct peer-to-peer connections in this configuration? 

**Options:**
1. Give peers network ports so they can accept connections and form a mesh
2. Accept star topology as correct for peers without public ports
3. Have gateway facilitate peer-to-peer connections somehow

What's the intended behavior?

[AI-assisted debugging and analysis]

sanity commented:
## Root Cause Analysis

I've identified the core issue. The current implementation has a logical flaw in how it handles connection acceptance.

### The Flawed Logic (Current Code)

```rust
// Current forward_conn() logic:
if num_connections > 0 {
    try_forward();
    if forward_failed {
        // Falls through to handle_unforwardable_connection
    }
}

// Falls through here if num_connections == 0 OR forward failed
handle_unforwardable_connection() {
    if accepted {
        return AwaitingConnectivity;  // Accept directly
    }
}
```

### Why This Is Wrong

**Problem**: `should_accept()` returns `true` for ANY node when `total_conn < min_connections` (default 25).

This means:
- Peer1 (non-gateway, 1 connection to gateway) will have `accepted == true` when peer2 tries to join
- When gateway forwards peer2's request to peer1, peer1 tries to forward it (has 1 connection)
- Forward fails (can only forward to gateway, which is already involved)
- Falls through to `handle_unforwardable_connection` with `accepted == true`
- **Incorrectly accepts peer2 directly instead of forwarding back to requester**

### The Fix

The issue is that `accepted` means "**this node is willing to have this connection**", NOT "**this node should accept the connection directly without forwarding**".

We need to distinguish between:
1. **Direct acceptance** (node has no choice, must accept) - only for gateways at 0 connections
2. **Reserve and forward** (node is willing, but should try to forward first) - normal case

I'm working on a fix that properly separates these cases.

[AI-assisted debugging and comment]

sanity commented:
@iduartgomez - I've been working on completing this PR and have identified what I believe is the final piece needed to make the gateway bootstrap work correctly.

## Current Status

✅ **Implemented:**
- Added `is_gateway` field to `ForwardParams` 
- Fixed `forward_conn()` to only accept directly when `num_connections == 0 && num_reserved == 0 && is_gateway && accepted`
- This prevents the race condition where multiple concurrent join attempts would all be accepted directly

❌ **Still Failing:**
- `test_three_node_network_connectivity` - mesh topology doesn't form
- Gateway accepts connections at transport level but never registers them in its ring
- Result: gateway can't respond to `FindOptimalPeer` requests because it doesn't know about connected peers

## Root Cause

When the gateway accepts in bootstrap mode, `forward_conn()` returns `AwaitingConnectivity` but **never sends a `CheckConnectivity` message**. 

Compare the two cases in `forward_conn()`:

**Forwarding case** (line 1094): 
```rust
let forward_msg = create_forward_message(/*...*/);  // Creates CheckConnectivity
network_bridge.send(&target_peer.peer, forward_msg).await?;  // Sends it
```

**Bootstrap case** (line 1054):
```rust
return Ok(Some(ConnectState::AwaitingConnectivity(connectivity_info)));  // No message sent!
```

The design intent from p2p_protoc.rs:839 says:
```rust
// IMPORTANT: Do NOT add connection to ring here!
// Connection should only be added after StartJoinReq is accepted
// via CheckConnectivity. This prevents the "already connected" bug
```

But `CheckConnectivity` never arrives for bootstrap connections, so they're never registered.

## Question

**For the gateway bootstrap case (0 connections, 0 reserved), when a peer initiates a direct outbound connection via `StartJoinReq`, should we:**

**A)** Immediately register the connection in the ring (bypassing the normal `CheckConnectivity` flow)?

**B)** Have `forward_conn()` send a `CheckConnectivity` message (to the gateway itself, or trigger it somehow)?

**C)** Something else?

The "already connected" concern doesn't apply to bootstrap since it's the very first connection. I'm leaning toward **Option A** - immediate registration for this special case.

Thoughts?

[AI-assisted debugging and comment]

iduartgomez commented:
Option A should be the most straightforward, document well where we do this case why we do this and reference to the logic in handshake_handler for future ref
